sotodlib/toast/workflows/sim_optical_pickup.py

```diff
@@ -0,0 +1,106 @@
+"""Simulated optical pickup from various sources.
+"""
+
+import numpy as np
+from astropy import units as u
+import toast
+import toast.ops
+
+from .. import ops as so_ops
+
+
+def setup_simulate_scan_synchronous_signal(operators):
+    """Add commandline args and operators for scan-synchronous signal.
+
+    Args:
+        operators (list):  The list of operators to extend.
+
+    Returns:
+        None
+
+    """
+    operators.append(toast.ops.SimScanSynchronousSignal(name="sim_sss", enabled=False))
+
+
+def simulate_scan_synchronous_signal(job, otherargs, runargs, data):
+    """Simulate scan-synchronous signal.
+
+    Args:
+        job (namespace):  The configured operators and templates for this job.
+        otherargs (namespace):  Other commandline arguments.
+        runargs (namespace):  Job related runtime parameters.
+        data (Data):  The data container.
+
+    Returns:
+        None
+
+    """
+    log = toast.utils.Logger.get()
+    timer = toast.timing.Timer()
+    timer.start()
+
+    # Configured operators for this job
+    job_ops = job.operators
+
+    if otherargs.realization is not None:
+        job_ops.sim_sss.realization = otherargs.realization
+
+    if job_ops.sim_sss.enabled:
+        job_ops.sim_sss.detector_pointing = job_ops.det_pointing_azel
+        log.info_rank(
+            "Running scan-synchronous signal simulation...", comm=data.comm.comm_world
+        )
+        job_ops.sim_sss.apply(data)
+        log.info_rank(
+            "Simulated Scan-synchronous signal", comm=data.comm.comm_world, timer=timer
+        )
+        job_ops.mem_count.prefix = "After simulating scan-synchronous signal"
+        job_ops.mem_count.apply(data)
+
+
+def setup_simulate_hwpss_signal(operators):
+    """Add commandline args and operators for HWP synchronous signal.
+
+    Args:
+        operators (list):  The list of operators to extend.
+
+    Returns:
+        None
+
+    """
+    operators.append(so_ops.SimHWPSS(name="sim_hwpss", enabled=False))
+
+
+def simulate_hwpss_signal(job, otherargs, runargs, data):
+    """Simulate HWP synchronous signal.
+
+    Args:
+        job (namespace):  The configured operators and templates for this job.
+        otherargs (namespace):  Other commandline arguments.
+        runargs (namespace):  Job related runtime parameters.
+        data (Data):  The data container.
+
+    Returns:
+        None
+
+    """
+    log = toast.utils.Logger.get()
+    timer = toast.timing.Timer()
+    timer.start()
+
+    # Configured operators for this job
+    job_ops = job.operators
+
+    if job_ops.sim_hwpss.enabled and job_ops.sim_ground.hwp_angle is not None:
+        job_ops.sim_hwpss.hwp_angle = job_ops.sim_ground.hwp_angle
+        job_ops.sim_hwpss.detector_pointing = job_ops.det_pointing_azel
+        job_ops.sim_hwpss.stokes_weights = job_ops.weights_azel
+        log.info_rank("Running HWPSS simulation...", comm=data.comm.comm_world)
```

Comment:
I like it when the workflow reports both when if begins an operation and when it ends. That makes it more apparent to a general user what is going on. However, since SO is acting as a model for these workflow tools, we should pick one convention and stick to it.

Reply:
Yes, my intention was to have a log message at the start of each thing and at the end with the timing info.  I'll go back through all the functions and make sure that pattern is followed.

Reply:
I implemented a decorator with logs the start / end time of each workflow function and the memory usage if it is enabled.  In some cases, when a workflow function is running several operators, I left the sub-timing information since it is useful.